### PR TITLE
Change how identities are shown in commit page

### PR DESCRIPTION
**Pull request by n0toose**
The username is now always shown next to full names. The commit
information should use the full name associated with the commit and the
signature information should use the full name associated with the
profile, provided that it exists.

This should remove any confusion if the author decides to use different
names under the same account at a later point in time.

In some cases, the full names shown in specific fields may differ from
the profile's full name, as some other metadata may be preferred
instead.

### CODE REVIEW

File: templates/repo/commit_page.tmpl [hunk 1]
```diff
@@ -201,7 +201,11 @@
 								<span class="ui text gt-mr-3">{{.locale.Tr "repo.commits.signed_by_untrusted_user_unmatched"}}:</span>
 							{{end}}
 							{{avatar $.Context .Verification.SigningUser 28 "gt-mr-3"}}
-							<a href="{{.Verification.SigningUser.HomeLink}}"><strong>{{.Verification.SigningUser.GetDisplayName}}</strong></a>
+							{{if .Verification.SigningUser.Fullname}}
```
Comment (jolheiser) [inline templates/repo/commit_page.tmpl]:
```suggestion
							{{if .Verification.SigningUser.FullName}}
```

**Comment by jolheiser** [comment]
The following is an image reference using my actual settings as I have them on various instances.
(It really drives home who the author of that commit is :joy:)

honk

---

I can see where this could be beneficial, but it almost seems like there should be a check on whether the names match, in which case it could just use one. i.e. `jolheiser (jolheiser)` doesn't look more clear, it just looks duplicated, however the GPG signature _does_ look a bit nicer because my full name is shown with my handle next to it. 
I can imagine that's also more helpful when people use less obvious handles than mine.

**Comment by silverwind** [comment]
We should unify the way author name is display. Sometimes it's username, sometimes it's full name, sometimes it's both. I'd say always display full name if present. If full name and username match, display only one time.

**Comment by wxiaoguang** [comment]
The easy way is to reuse the template: https://github.com/go-gitea/gitea/blob/main/templates/repo/search_name.tmpl

At the moment the Gitea's name displaying is not clear / inconsistent, that's a historical problem.

If you think `search_name.tmpl` could help, then it could be used here IMO (although the template name doesn't look good 🤣 )

If you prefer your current approach, it could also look good.

(I think I will do some "name displaying" refactoring in following PRs .......  😄 )

**Comment by wxiaoguang** [comment]
> Can their username be discovered somehow?

I haven't understood the question "discover the username".

If I understand correctly, if the `.Verification.CommittingUser` and `.Commit.Committer` are `User` models, then the `.Name` is the username?

**Comment by n0toose** [comment]
> If I understand correctly, if the .Verification.CommittingUser and .Commit.Committer are User models, then the .Name is the username?

This is the part that has managed to confuse me very much, `.Name` seems to be the full name (as presented in their profile).

**Comment by wxiaoguang** [comment]
I think I somewhat understand these variables now. The details are in `routers/web/repo/commit.go`, `func Diff(ctx *context.Context)`:

* `Author`: it is a `User` model
* `Commit.Author`: it is `modules/git/Commit` and `modules/git/Signature`:
    * `Name string` // Name represents a person name. It is an arbitrary string. It's the name from git commit
* `Verification.SigningUser`: it is `modules/git/CommitVerification` and `User` model

**Comment by n0toose** [comment]
I just glanced over the code, thank you for the hints and the help here. So, just to be sure, is it true that we can't correlate committers and their respective on-platform usernames as things are right now?

**Comment by wxiaoguang** [comment]
> is it true that we can't correlate committers and their respective on-platform usernames as things are right now?

If the commit user can be found in Gitea's database (hmm, many more details, eg: by email, by key) as a site user, then these two "names" are correlated. Otherwise, not correlated because no clue to correlate them.

**Comment by n0toose** [comment]
> If the commit user can be found in Gitea's database (hmm, many more details, eg: by email, by key) as a site user, then these two "names" are correlated.

Hi, I am not sure if I can implement all that by myself because of a personal lack of experience and a hesitance of dealing with PII like this...

**Comment by wxiaoguang** [comment]
I think this PR could help you:  Add a helper function dumpVar to help debugging and development #24262 (at least, we can know what variable type it is)

**Comment by n0toose** [comment]
Most of the things I typed out earlier regarding "committers being completely ignored" were wrong. `.Verification.CommittingUser` made me think that this parameter had something to do with PGP signatures, which was not the case.

I managed to figure out how the template works in greater detail by playing around with the code, breaking things, then undoing those things only to realize that the features that I described as "missing" existed in the first place. Anyways, this latest push is on top of the latest revision of the tree and ensures that Committers will follow the same format: Full Name As Stored In Examined Resource (Username) *(excluding PGP key name, which uses Forgejo profile name instead)*

**Comment by n0toose** [comment]
image

File: templates/repo/commit_page.tmpl [hunk 2]
```diff
@@ -201,7 +209,11 @@
 								<span class="ui text gt-mr-3">{{.locale.Tr "repo.commits.signed_by_untrusted_user_unmatched"}}:</span>
 							{{end}}
 							{{avatar $.Context .Verification.SigningUser 28 "gt-mr-3"}}
-							<a href="{{.Verification.SigningUser.HomeLink}}"><strong>{{.Verification.SigningUser.GetDisplayName}}</strong></a>
+							{{if .Verification.SigningUser.FullName}}
+								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> {{.Verification.SigningUser.FullName}} ({{.Verification.SigningUser.GetDisplayName}})</strong></a>
```
Comment (wxiaoguang) [inline templates/repo/commit_page.tmpl:213]:
```suggestion
							{{if and DefaultShowFullName .Verification.SigningUser.FullName}}
								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> {{.Verification.SigningUser.Name}} ({{.Verification.SigningUser.FullName}})</strong></a>
```

I guess it should be like this?

File: templates/repo/commit_page.tmpl [hunk 3]
```diff
@@ -201,7 +209,11 @@
...
+							{{else}}
+								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> ({{.Verification.SigningUser.GetDisplayName}})</strong></a>
+							{{end}}
```
Comment (wxiaoguang) [inline templates/repo/commit_page.tmpl]:
```suggestion
							{{if and DefaultShowFullName .Verification.SigningUser.FullName}}
								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> {{.Verification.SigningUser.Name}} ({{.Verification.SigningUser.FullName}})</strong></a>
							{{else}}
								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> ({{.Verification.SigningUser.Name}})</strong></a>
							{{end}}
```


I guess it should be like this?

**Review by wxiaoguang — COMMENTED** [review]
Generally L-G-T-M.

Could we have some inline comments about the types of each variable? eg:

```

{{/* .Author: User model */}}

{{/* .Commit.Author:  modules/git/Commit, modules/git/Signature */}}

{{/* .Verification.SigningUser:  modules/git/CommitVerification, User model */}}

```


It would help future developers a lot.

File: templates/repo/commit_page.tmpl (hunk 3 above)
Comment (lunny) [reply templates/repo/commit_page.tmpl]: FullName maybe blank?

File: templates/repo/commit_page.tmpl [hunk 4]
```diff
@@ -201,7 +214,12 @@
 								<span class="ui text gt-mr-3">{{.locale.Tr "repo.commits.signed_by_untrusted_user_unmatched"}}:</span>
 							{{end}}
 							{{avatar $.Context .Verification.SigningUser 28 "gt-mr-3"}}
-							<a href="{{.Verification.SigningUser.HomeLink}}"><strong>{{.Verification.SigningUser.GetDisplayName}}</strong></a>
+							{{if .Verification.SigningUser.FullName}}
+								{{/* Use the same scheme as the one used in commits. */}}
+								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> {{.Verification.SigningUser.FullName}} ({{.Verification.SigningUser.GetDisplayName}})</strong></a>
```
Comment (wxiaoguang) [inline templates/repo/commit_page.tmpl]: I guess it could be like this? `Name (FullName)` or `FullName (Name)`.  Update:`FullName` might be empty.

```suggestion
								<a href="{{.Verification.SigningUser.HomeLink}}"><strong> {{.Verification.SigningUser.Name}} ({{.Verification.SigningUser.FullName}})</strong></a>
```


Otherwise, GetDisplayName sometimes returns "Name", sometimes returns "FullName".

Overall LGTM while there is still a question in my mind: when to use `user.Name` , when to use `user.FullName` ? (see another comment)
 
----

Update, quote another review comment:

* https://github.com/go-gitea/gitea/pull/24087#discussion_r1176423185

The `FullName` might be empty, in this case, `GetDisplayName` also returns `Name`.

File: templates/repo/commit_page.tmpl [hunk 5]
```diff
@@ -145,10 +145,18 @@
...
 						{{avatar $.Context .Author 28 "gt-mr-3"}}
-						{{if .Author.FullName}}
-							<a href="{{.Author.HomeLink}}"><strong>{{.Author.FullName}}</strong></a>
+						{{if .Author.Name}}
+							{{if eq .Commit.Author.Name .Author.Name}}
+								<a href="{{.Author.HomeLink}}"><strong>{{.Commit.Author.Name}}</strong></a>
+							{{else}}
+								{{/* If the account name and username don't match, use an alternative "also known as" format */}}
+								<a href="{{.Author.HomeLink}}"><strong>{{.Commit.Author.Name}} ({{.Author.Name}})</strong></a>
+							{{end}}
```
Comment (wxiaoguang) [inline templates/repo/commit_page.tmpl]: (follow my previous comment and question)

Here the logic is changed: before `{{if .Author.FullName}}` , now `{{if .Author.Name}}`, while I think `.Author.Name` should always exist, it is the username, so this `if` is a no-op?


And, here it never shows user's full name, is it done on purpose?  (quote my previous question: when to use user.Name , when to use user.FullName)

----

Dismiss, I haven't had enough understanding for this problem.

File: templates/repo/commit_page.tmpl (hunk 4 above)
Comment (n0toose) [reply templates/repo/commit_page.tmpl]: `GetDisplayName` presumably respects `DefaultShowFullName`, correct?

Comment (wxiaoguang) [reply templates/repo/commit_page.tmpl]: IIRC it works like this:

```go
if DefaultShowFullName && FullName != "" {
    return FullName
} else {
    return Name
}
```

Comment (n0toose) [reply templates/repo/commit_page.tmpl]: Okay, I didn't know about this. Thanks, I'll try to incorporate this better, but the thing is that I do show the full name and the username in both cases on purpose sometimes. I will have to reevaluate the logic of what I did, look at how I could incorporate DefaultShowFullName and then remove redundant cases using GetDisplayName. (and then get back to you)

Comment (wxiaoguang) [reply templates/repo/commit_page.tmpl]: I didn't mean following `DefaultShowFullName` is a must. As long as the new display logic is complete, consistent and reasonable, then it's good 😄

**Comment by n0toose** [comment]
@wxiaoguang Apart from the comment that I haven't addressed at all yet (the logic behind this PR is extremely hard to follow if you haven't extensively worked with it over the past 24 hours to me), do you have any suggestions as to how my approach could be done in a cleaner manner?

**Comment by wxiaoguang** [comment]
> do you have any suggestions as to how my approach could be done in a cleaner manner?

Sorry I have no idea at the moment (I haven't got enough time and I haven't had enough understanding for this problem yet). I might take a look in the future if I have enough time but I can't guarantee at the moment.

So I hide my reviews. If other reviewers have ideas or approve, feel free to continue. Thank you.

**Comment by wxiaoguang** [comment]
I tried to read the code again, I think it could be like this:  fix name display #1

**Comment by n0toose** [comment]
Blindly merged for now, I'll get back to this at a later point (this is still on my radar).

**Comment by wxiaoguang** [comment]
Found some problems in my code, so added a new commit to fix them, and added some comments, I think the "user name" will be rendered in the following formats now:

* If ShowFulName== false:
    * UserName   // if the CommiterName == UserName
    * CommiterName (UserName)   // if the CommiterName != UserName
* If ShowFulName== true:
    * UserName   // if the CommiterName == UserName and no FullName
    * UserName (FullName)  // if the CommiterName == UserName and FullName exists
    * CommiterName (UserName)   // if the CommiterName != UserName and no FullName
    * CommiterName (UserName / FullName)   // if the CommiterName != UserName and FullName exists

**Comment by n0toose** [comment]
Really sorry for taking up your time with this earlier, but I think that I lack the time and concentration to actually solve this problem properly right now. I think that a better approach would be modifying the `GetDisplayName()` function properly instead of relying on a lot of conditionals in the web UI itself, which increases the overhead and is plain confusing.

If I am to pick this up again (which I remain interested in, just not interested enough to do this right now), I will do so in a new PR (maybe with a much more limited scope for starters). For now, I'll leave the space for someone else to do this, just in case this person appears.
